content/en/blog/_posts/2021-XX-XX-writing-a-controller-for-pod-labels.md

```diff
@@ -0,0 +1,473 @@
+
+I recently worked on a project where we needed to create a Service that routed
+traffic to a specific Pod in a ReplicaSet. The problem is that a Service can
+only select Pods by label, and all pods in a ReplicaSet have the same labels.
+There are two ways to solve this problem:
+
+1. Create a Service without a selector and manage the Endpoints or
+   EndpointSlices for that Service directly. We would need to write a custom
+   controller to insert our Pod's IP address into those resources.
+2. Add a label to the Pod a unique value. We could then use this label in our
+   Service's selector. Again, we would need to write a custom controller to add
+   this label.
+
+A controller is a control-loop that tracks one or more Kubernetes resource
+types. The controller from option n°2 above only needs to track Pods, which
+makes it simpler to implement. This is the option we are going to walk through
+by writing a Kubernetes controller that adds a `pod-name` label to our Pods.
+
+StatefulSets [do this natively][statefulset-pod-name-label] by adding a
+`pod-name` label to each Pod in the set. But what if we don't want to or can't
+use StatefulSets?
+
+We rarely create Pods directly; most often, we use a Deployment, ReplicaSet, or
+another high-level resource. We can specify labels to add to each Pod in the
+PodSpec, but not with dynamic values, so no way to replicate a StatefulSet's
+`pod-name` label.
+
+We tried using a [mutating admission webhook][mutating-admission-webhook]. When
+anyone creates a Pod, the webhook patches the Pod with a label containing the
+Pod's name. Disappointingly, this does not work: not all Pods have a name before
+being created. For instance, when the ReplicaSet controller creates a Pod, it
+sends a `namePrefix` to the Kubernetes API server and not a `name`. The API
+server generates a unique name before persisting the new Pod to etcd, but only
+after calling our admission webhook. So in most cases, we can't know a Pod's
+name with a mutating webhook.
+
+Once a Pod exists in the Kubernetes API, it is mostly immutable, but we can
+still add a label. We can even do so from the command line:
+
+```bash
+kubectl label my-pod my-label-key=my-label-value
+```
+
+We need to watch for changes to any Pods in the Kubernetes API and add the label
+we want. Rather than do this manually, we are going to write a controller that
+does it for us.
+
+## Bootstrapping a controller with the Operator SDK
+
+An operator is a set of controllers that, when combined, serve to manage an
+application running inside Kubernetes. A controller is a reconciliation loop
+that reads the desired state of a resource from the Kubernetes API and takes
+action to bring the cluster's actual state closer to the desired state.
+
+In order to write this controller as quickly as possible, we are going to use
+the Operator SDK. If you don't have it installed, follow the
+[official documentation][operator-sdk-installation].
+
+```terminal
+$ operator-sdk version
```

Comment:
```suggestion
operator-sdk version
```
https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt

Reply:
You can ignore this suggestion.